apiserver/facades/client/application/application_test.go

```diff
@@ -2493,53 +2532,108 @@ func (s *applicationSuite) setupApplicationExpose(c *gc.C) {
 		apps[i] = s.AddTestingApplication(c, name, charm)
 		c.Assert(apps[i].IsExposed(), jc.IsFalse)
 	}
-	err = apps[1].SetExposed(nil)
+	err = apps[1].MergeExposeSettings(nil)
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(apps[1].IsExposed(), jc.IsTrue)
 }
 
 var applicationExposeTests = []struct {
-	about       string
-	application string
-	err         string
-	exposed     bool
+	about                 string
+	application           string
+	exposedEndpointParams map[string]params.ExposedEndpoint
+	//
+	expExposed          bool
+	expExposedEndpoints map[string]state.ExposedEndpoint
+	expErr              string
 }{
 	{
 		about:       "unknown application name",
 		application: "unknown-application",
-		err:         `application "unknown-application" not found`,
+		expErr:      `application "unknown-application" not found`,
 	},
 	{
-		about:       "expose a application",
+		about:       "expose all endpoints of an application ",
 		application: "dummy-application",
-		exposed:     true,
+		expExposed:  true,
+		expExposedEndpoints: map[string]state.ExposedEndpoint{
+			"": {
+				ExposeToCIDRs: []string{"0.0.0.0/0"},
+			},
+		},
 	},
 	{
 		about:       "expose an already exposed application",
 		application: "exposed-application",
-		exposed:     true,
+		expExposed:  true,
+		expExposedEndpoints: map[string]state.ExposedEndpoint{
+			"": {
+				ExposeToCIDRs: []string{"0.0.0.0/0"},
+			},
+		},
+	},
+	{
+		about:       "unknown endpoint name in expose parameters",
+		application: "dummy-application",
+		exposedEndpointParams: map[string]params.ExposedEndpoint{
+			"bogus": {},
+		},
+		expErr: `endpoint "bogus" not found`,
+	},
+	{
+		about:       "unknown space name in expose parameters",
+		application: "dummy-application",
+		exposedEndpointParams: map[string]params.ExposedEndpoint{
+			"": {
+				ExposeToSpaces: []string{"invaders"},
+			},
+		},
+		expErr: `space "invaders" not found`,
```

Comment:
It's good to know that all that R&D effort for modulating shield frequency harmonics has finally paid off ;-) 👾

Reply:
:space_invader: :gun: